File: gazavba/backend/models/Status.js
const { db } = require('../config/database');
const { v4: uuidv4 } = require('uuid');

class Status {
  static async create(statusData) {
    const { userId, type = 'text', content, mediaUrl, expiresAt } = statusData;
    const id = uuidv4();
    
    return new Promise((resolve, reject) => {
      db.run(
        `INSERT INTO statuses (id, userId, type, content, mediaUrl, expiresAt) VALUES (?, ?, ?, ?, ?, ?)`,
        [id, userId, type, content, mediaUrl, expiresAt],
        function(err) {
          if (err) reject(err);
          else resolve({ id, ...statusData, createdAt: new Date() });
        }
      );
    });
  }

  static async getByUserId(userId) {
    return new Promise((resolve, reject) => {
      db.all(
        `SELECT s.*, u.name as userName, u.avatar as userAvatar 
         FROM statuses s 
         JOIN users u ON s.userId = u.id 
         WHERE s.userId = ? AND (s.expiresAt IS NULL OR s.expiresAt > datetime('now')) 
         ORDER BY s.createdAt DESC`,
        [userId],
        (err, rows) => {
          if (err) reject(err);
          else resolve(rows);
        }
      );
    });
  }

  static async getAll(userId) {
    return new Promise((resolve, reject) => {
      db.all(
        `SELECT s.*, u.name as userName, u.avatar as userAvatar,
         (SELECT COUNT(*) FROM status_views sv WHERE sv.statusId = s.id) as viewCount,
         (SELECT COUNT(*) FROM status_views sv WHERE sv.statusId = s.id AND sv.viewerId = ?) as hasViewed
         FROM statuses s 
         JOIN users u ON s.userId = u.id 
         WHERE (s.expiresAt IS NULL OR s.expiresAt > datetime('now')) 
         ORDER BY s.createdAt DESC`,
        [userId],
        (err, rows) => {
          if (err) reject(err);
          else resolve(rows);
        }
      );
    });
  }

  static async markAsViewed(statusId, viewerId) {
    return new Promise((resolve, reject) => {
      db.run(
        `INSERT OR IGNORE INTO status_views (id, statusId, viewerId) VALUES (?, ?, ?)`,
        [uuidv4(), statusId, viewerId],
        function(err) {
          if (err) reject(err);
          else resolve();
        }
      );
    });
  }

  static async getViewers(statusId) {
    return new Promise((resolve, reject) => {
      db.all(
        `SELECT u.*, sv.viewedAt FROM users u 
         JOIN status_views sv ON u.id = sv.viewerId 
         WHERE sv.statusId = ? 
         ORDER BY sv.viewedAt DESC`,
        [statusId],
        (err, rows) => {
          if (err) reject(err);
          else resolve(rows);
        }
      );
    });
  }

  static async delete(id) {
    return new Promise((resolve, reject) => {
      db.run(`DELETE FROM statuses WHERE id = ?`, [id], function(err) {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  static async getUnseenCount(userId) {
    return new Promise((resolve, reject) => {
      db.get(
        `SELECT COUNT(*) as count FROM statuses s 
         WHERE s.userId != ? AND (s.expiresAt IS NULL OR s.expiresAt > datetime('now'))
         AND NOT EXISTS (SELECT 1 FROM status_views sv WHERE sv.statusId = s.id AND sv.viewerId = ?)`,
        [userId, userId],
        (err, row) => {
          if (err) reject(err);
          else resolve(row.count);
        }
      );
    });
  }
}

module.exports = Status;

File: lib/core/models/status.dart
import '../utils/url_utils.dart';

class Status {
  const Status({
    required this.id,
    required this.userId,
    required this.createdAt,
    required this.expiresAt,
    this.durationSeconds,
    this.content,
    this.mediaUrl,
    this.userName,
    this.userAvatar,
    this.viewCount = 0,
    this.hasViewed = false,
  });

  factory Status.fromJson(Map<String, dynamic> json) {
    return Status(
      id: json['id'].toString(),
      userId: json['userId']?.toString() ?? json['user_id']?.toString() ?? '',
      createdAt: DateTime.tryParse(json['createdAt'] as String? ??
              json['created_at'] as String? ??
              '') ??
          DateTime.now(),
      expiresAt: DateTime.tryParse(json['expiresAt'] as String? ??
              json['expires_at'] as String? ??
              '') ??
          DateTime.now().add(const Duration(hours: 24)),
      durationSeconds: _asInt(json['durationSeconds'] ?? json['duration']),
      content: json['content'] as String?,
      mediaUrl: UrlUtils.resolveMediaUrl(
        json['mediaUrl'] as String? ?? json['media_url'] as String?,
      ),
      userName: json['userName'] as String? ?? json['user_name'] as String?,
      userAvatar: UrlUtils.resolveMediaUrl(
        json['userAvatar'] as String? ?? json['user_avatar'] as String?,
      ),
      viewCount: _asInt(json['viewCount'] ?? json['view_count']),
      hasViewed: (json['hasViewed'] ?? json['has_viewed']) == true ||
          (json['hasViewed'] == 1 || json['has_viewed'] == 1),
    );
  }

  final String id;
  final String userId;
  final DateTime createdAt;
  final DateTime expiresAt;
  final int? durationSeconds;
  final String? content;
  final String? mediaUrl;
  final String? userName;
  final String? userAvatar;
  final int viewCount;
  final bool hasViewed;

  Status copyWith({
    bool? hasViewed,
    int? viewCount,
    int? durationSeconds,
  }) {
    return Status(
      id: id,
      userId: userId,
      createdAt: createdAt,
      expiresAt: expiresAt,
      durationSeconds: durationSeconds ?? this.durationSeconds,
      content: content,
      mediaUrl: mediaUrl,
      userName: userName,
      userAvatar: userAvatar,
      viewCount: viewCount ?? this.viewCount,
      hasViewed: hasViewed ?? this.hasViewed,
    );
  }
}

int _asInt(dynamic value) {
  if (value is int) return value;
  if (value is double) return value.toInt();
  if (value is String) return int.tryParse(value) ?? 0;
  return 0;
}
